### PR TITLE
Sammcdo/2/turret

### PR DESCRIPTION
**Pull request by sammcdo**

### CODE REVIEW

File: src/main/java/frc/robot/Constants.java [hunk 1]
```diff
@@ -19,7 +19,20 @@
     public static final int CAN_ID_BACK_RIGHT_DRIVE_MOTOR = 10;
     public static final int USB_PORT_DRIVER_CONTROLLER = 0;
     public static final double DRIVEBASE_DEADBAND = 0.015; // anything under 1.5% of full joystick away from 0 should be considered 0
+    public static final double TURRET_DEADBAND = 0.015; // anything under 1.5% of full joystick away from 0 should be considered 0
     public static final double DRIVEBASE_TURN_SLEW_LIMIT = 2.0; // 0 to 200% in one second. 0 to full in 1/2 second.
     public static final double DRIVEBASE_FWD_BACK_SLEW_LIMIT = 1; // 0 to 100% in one second.
+    
+    /*Turret Constants*/
+    public static final double ENCODER_COUNTS_PER_TURRET_NEO_REVOLUTION = 1;
+    public static final double NEO_REVOLUTIONS_PER_TURRET_REVOLUTION = 5.23 * 150/16;//2.89 * 190 / 16;
```
Comment (sammcdo) [inline src/main/java/frc/robot/Constants.java:28]: change this to new robot gearing

File: src/main/java/frc/robot/Constants.java [hunk 2]
```diff
@@ -19,7 +19,20 @@
...
+    public static final double ENCODER_COUNTS_PER_TURRET_DEGREE = ENCODER_COUNTS_PER_TURRET_NEO_REVOLUTION * NEO_REVOLUTIONS_PER_TURRET_REVOLUTION / 360;
+    public static final int CAN_ID_TURRET = 30;
+    public static final double TURRET_SPEED = -0.1;
+    public static final double TURRET_REVERSE_SOFT_LIMIT_DEGREES = 45;
```
Comment (sammcdo) [inline src/main/java/frc/robot/Constants.java:32]: is this used anymore?

File: src/main/java/frc/robot/RobotContainer.java [hunk 3]
```diff
@@ -20,9 +28,20 @@
 public class RobotContainer {
   // The robot's subsystems and commands are defined here...
   /** The container for the robot. Contains subsystems, OI devices, and commands. */
-  private XboxController driverController;
+  private XboxController turretController;
```
Comment (sammcdo) [inline src/main/java/frc/robot/RobotContainer.java:31]: this needs to be renamed to driverController

File: src/main/java/frc/robot/commands/TurretHomingC.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+// the WPILib BSD license file in the root directory of this project.
+
+package frc.robot.commands;
+
+import edu.wpi.first.wpilibj2.command.CommandBase;
+import frc.robot.Constants;
+import frc.robot.subsystems.TurretS;
+
+public class TurretHomingC extends CommandBase {
+  /** Creates a new TurretHomingC. */
```
Comment (sammcdo) [inline src/main/java/frc/robot/commands/TurretHomingC.java:12]: can you add javadoc to explain what this does?

File: src/main/java/frc/robot/commands/TurretManualC.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+import edu.wpi.first.wpilibj2.command.CommandBase;
+import frc.robot.subsystems.TurretS;
+/**
+ * control the turret
+ */
+public class TurretManualC extends CommandBase {
+  private XboxController turretController;
+  private TurretS turretS;
+
+  /** Creates a new TurretManualC. */
```
Comment (sammcdo) [inline src/main/java/frc/robot/commands/TurretManualC.java:17]: Can you add some javadoc to explain what this does?

File: src/main/java/frc/robot/commands/TurretTurnC.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import edu.wpi.first.wpilibj2.command.CommandBase;
+import frc.robot.subsystems.TurretS;
+
+public class TurretTurnC extends CommandBase {
+  private TurretS turret;
+  private double angle;
+  
+  /** Creates a new TurretTurnC. */
```
Comment (sammcdo) [inline src/main/java/frc/robot/commands/TurretTurnC.java:15]: same here, some javadoc would be awesome

File: src/main/java/frc/robot/commands/TurretTurnC.java [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+  // Called when the command is initially scheduled.
+  @Override
+  public void initialize() {}
+
+  // Called every time the scheduler runs while the command is scheduled.
+  @Override
+  public void execute() {
+    //double difference = 40 - turret.getEncoderCounts(); 
+    //SmartDashboard.putNumber("turret error", difference);
+    //turret.turnSpeed((difference * 0.0025) + (0.02 * Math.signum(difference)));
```
Comment (sammcdo) [inline src/main/java/frc/robot/commands/TurretTurnC.java:32]: do we need these lines still?

File: src/main/java/frc/robot/subsystems/TurretS.java [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+ * @author Benjamin Su
+ * @author Noah Kim
+ */
+public class TurretS extends SubsystemBase implements Loggable {
+  private CANSparkMax sparkMax = new CANSparkMax(Constants.CAN_ID_TURRET, MotorType.kBrushless);
+  private RelativeEncoder sparkMaxEncoder = sparkMax.getEncoder(SparkMaxRelativeEncoder.Type.kHallSensor, 42);
+  private DigitalInput limitSwitch = new DigitalInput(Constants.LIMIT_SWITCH_PORT_NUMBER);
+  private PIDController pid = new PIDController(0.0025, 0, 0);
+
+  /** Creates a new TurretS. */
```
Comment (sammcdo) [inline src/main/java/frc/robot/subsystems/TurretS.java:34]: can this be expanded? maybe put a description and authors

**Review by sammcdo — COMMENTED** [review]
Looking great! If you could run a vs code format for each file and check out the comments it should be ready to score

File: src/main/java/frc/robot/subsystems/TurretS.java (hunk 8 above)
Comment (shueja-personal) [reply src/main/java/frc/robot/subsystems/TurretS.java:34]: See the class javadoc.

File: src/main/java/frc/robot/Constants.java [hunk 9]
```diff
@@ -19,7 +19,20 @@
     public static final int CAN_ID_BACK_RIGHT_DRIVE_MOTOR = 10;
     public static final int USB_PORT_DRIVER_CONTROLLER = 0;
     public static final double DRIVEBASE_DEADBAND = 0.015; // anything under 1.5% of full joystick away from 0 should be considered 0
+    public static final double TURRET_DEADBAND = 0.015; // anything under 1.5% of full joystick away from 0 should be considered 0
     public static final double DRIVEBASE_TURN_SLEW_LIMIT = 2.0; // 0 to 200% in one second. 0 to full in 1/2 second.
     public static final double DRIVEBASE_FWD_BACK_SLEW_LIMIT = 1; // 0 to 100% in one second.
+    
+    /*Turret Constants*/
+    public static final double ENCODER_COUNTS_PER_TURRET_NEO_REVOLUTION = 1;
+    public static final double NEO_REVOLUTIONS_PER_TURRET_REVOLUTION = 2.89 * 188 / 16;
```
Comment (BroDudeGuyMan) [inline src/main/java/frc/robot/Constants.java:28]: should be 188/21

**Pull request merged**